dotnet/src/Microsoft.Agents.AI.Hosting.OpenAI/Responses/Common/AsyncEnumerableExtensions.cs

```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Agents.AI.Hosting.OpenAI.Responses.Common;
+
+/// <summary>
+/// Extension methods for IAsyncEnumerable.
+/// </summary>
+internal static class AsyncEnumerableExtensions
+{
+    /// <summary>
+    /// Chunks the source based on when items change.
+    /// </summary>
+    /// <typeparam name="TSource">The source item type.</typeparam>
+    /// <param name="source">The async enumerable source.</param>
+    /// <param name="isChanged">Function to determine if items changed.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>Chunked async enumerable.</returns>
+    public static IAsyncEnumerable<IAsyncEnumerable<TSource>> ChunkOnChangeAsync<TSource>(
+        this IAsyncEnumerable<TSource> source,
+        Func<TSource?, TSource?, bool>? isChanged = null,
+        CancellationToken cancellationToken = default)
+    {
+        var c = isChanged == null
+            ? EqualityComparer<TSource>.Default
+            : EqualityComparer<TSource>.Create((x, y) => !isChanged(x, y), t => t?.GetHashCode() ?? 0);
+
+        return source.ChunkByKeyAsync(x => x, c, cancellationToken);
+    }
+
+    /// <summary>
+    /// Chunks the source by key selector.
+    /// </summary>
+    /// <typeparam name="TSource">The source item type.</typeparam>
+    /// <typeparam name="TKey">The key type.</typeparam>
+    /// <param name="source">The async enumerable source.</param>
+    /// <param name="keySelector">Function to select the key.</param>
+    /// <param name="comparer">Optional key comparer.</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>Chunked async enumerable by key.</returns>
+    public static async IAsyncEnumerable<IAsyncEnumerable<TSource>> ChunkByKeyAsync<TSource, TKey>(
+        this IAsyncEnumerable<TSource> source,
+        Func<TSource, TKey> keySelector,
+        IEqualityComparer<TKey>? comparer = null,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        comparer ??= EqualityComparer<TKey>.Default;
+
+        var e = source.GetAsyncEnumerator(cancellationToken);
```

Comment:
If this were a general purpose routine, we'd want to make sure this was disposed. If we know source will never be anything that needs disposal, we could get away without it, but in general it's still better to clean up.